feat: add evergreen staking to action center

premiumjibles opened the pull request:
## Description

This implements the action centre notification for evergreen staking

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #9814

## Risk

> High Risk PRs Require 2 approvals

Low risk, this implements the notification system of a somewhat peripheral feature (fox farming). It's also behind a feature flag (action center)

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

To test this. You must buy some WETH/FOX token **on ethereum**. Then navigate to the fox ecosystem page, scroll to the bottom and select deposit for fox farming. This PR implements the notification system for that deposit.

To test pending states you must close the pending modal for the fox farming deposit.

* Test that the toast shows up when the transaction succeeds or fails
* Test that the action centre  shows the pending and complete transaction
* Test that the click through links for contract address and transaction works

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

- [] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)
https://jam.dev/c/768a91e2-942a-430f-80d4-0a343cbc1077

premiumjibles committed:
feat: add evergreen staking to action center

gomesalexandre commented on src/components/Layout/Header/ActionCenter/components/EvergreenDepositActionCard.tsx:
```diff
@@ -0,0 +1,169 @@
+                <Stack spacing={3}>
+                  {txLink && (
+                    <Button as={Link} href={txLink} isExternal size='sm' width='full'>
+                      View Transaction
```
Translation

gomesalexandre commented on src/features/defi/providers/fox-farming/components/FoxFarmingManager/Deposit/components/Confirm.tsx:
```diff
@@ -128,6 +134,31 @@ export const Confirm: React.FC<StepComponentProps & { accountId: AccountId | und
       dispatch({ type: FoxFarmingDepositActionType.SET_TXID, payload: txid })
       onOngoingFarmingTxIdChange(txid, contractAddress)
+
+      if (contractAddress === ETH_FOX_STAKING_EVERGREEN_CONTRACT && asset && accountId) {
```
Why for Evergreen only?

gomesalexandre replied:
re: this one, since this is already in the FOX Farming deposit DeFi modal, I think it doesn't hurt to remove the check altogether, but not required as part of this PR if you want to keep it for now

gomesalexandre commented on src/hooks/useActionCenterSubscriber/useEvergreenDepositActionSubscriber.tsx:
```diff
@@ -0,0 +1,85 @@
+
+      const accountAddress = fromAccountId(accountId).account
+      const serializedTxIndex = serializeTxIndex(accountId, stakeTxHash, accountAddress)
+      const transaction = txById[serializedTxIndex]
```
💜

gomesalexandre commented on src/hooks/useActionCenterSubscriber/useEvergreenDepositActionSubscriber.tsx:
```diff
@@ -0,0 +1,85 @@
+    }
+  }, [isDrawerOpen, toast, previousIsDrawerOpen])
+
+  const pendingEvergreenActions = useAppSelector(selectPendingEvergreenDepositActions)
```
That's absolutely fine for the time being, but this won't scale, we should probably think one more level of abstraction here and have a hook that does generalised Txs introspection (it itself *could* leverage multiple lookup hooks)

gomesalexandre replied:
Also re: the question of inserting from where staking happens, and updating here, not sure about this one either.

It doesn't hurt to, but feels like it adds overhead, we could simply select Txs filtered by `to` being RFOX contracts (see `maybeRefetchOpportunities` in `src/hooks/useTransactionsSubscriber.ts` for an example of where we do that)

gomesalexandre replied:
Closing spewy wall of text

gomesalexandre commented on src/hooks/useActionCenterSubscriber/useEvergreenDepositActionSubscriber.tsx:
```diff
@@ -0,0 +1,85 @@
+
+      if (
+        !transaction ||
+        (transaction.status !== TxStatus.Failed && transaction.status !== TxStatus.Confirmed)
```
q: if we're going with handling all Tx statuses, do we even need this? Really only unknown should be a noop, and tbh I think we *should* be able to clean this one up from the app, that should never happen (i.e if we have a Tx, it's either pending, success or failed)

gomesalexandre commented on src/components/Layout/Header/ActionCenter/components/EvergreenDepositActionCard.tsx:
```diff
@@ -0,0 +1,169 @@
+    action.evergreenDepositMetadata
+
+  const feeAsset = useAppSelector(state =>
+    selectAssetById(state, lpAsset.chainId === 'eip155:1' ? 'eip155:1/slip44:60' : ''),
```
ser

gomesalexandre reviewed (commented):
Code LGTM as far as doing the thing for Evergreen deposits specifically - though few concerns here both from code and product perspective:

### Generalisation

- Not sure about the granularity of a FOX staking deposit action, IMO this should be generalised to Tx actions, or we'd have to add as many actions as there are Tx domains which doesn't sound nice and generalised at all. From a conceptual level, all this needs is an AssetId (for icons/precision), an amount, and a copy, nothing specific to FOX staking
Wondering if we want to get https://github.com/shapeshift/web/pull/9845 in first and leverage Tx transactions?


### Product

- I know this is in the Figma mocks, but the `to <foxStakingContractAddress>` part doesn't look user-friendly at all. I believe it's a relic from research and not ready to handoff, let's triple-check with @shapeshift/product but I don't think we want this
- Not generalised to other FOX staking opportunities. Totally understand why this is done like that since the ticket was all but clear, but we should ensure that generalizes to all "FOX Farming" (a.k.a `DefiProvider.EthFoxStaking`) opportunities and not the Evergreen one only, as Evergreen is only one of many:

<img width="1368" alt="Image" src="https://github.com/user-attachments/assets/e47999b8-5280-4671-9866-7d776ca02e3f" />

- The issue only mentions staking, but there is no mention of unstaking there and no matching unstaking issue, so my understanding is that unstake should also be done as part of it

gomesalexandre commented:
Also @premiumjibles can we uncheck this from description pls

<img width="593" alt="image" src="https://github.com/user-attachments/assets/33216f96-36bd-4046-b3ef-1c2112173c70" />

Action center *is* live in prod

<img width="1607" alt="image" src="https://github.com/user-attachments/assets/b0a5f5f2-00be-4209-8910-a516f1118033" />

premiumjibles commented:
> I know this is in the Figma mocks, but the to <foxStakingContractAddress> part doesn't look user-friendly at all. I believe it's a relic from research and not ready to handoff, let's triple-check with @shapeshift/product but I don't think we want this

Agree with this @gomesalexandre , i've removed

premiumjibles committed:
Action PR feedback

gomesalexandre commented on src/hooks/useActionCenterSubscribers/useEvergreenDepositActionSubscriber.tsx:
```diff
@@ -0,0 +1,79 @@
+  }, [isDrawerOpen, toast, previousIsDrawerOpen])
+
+  const pendingEvergreenActions = useAppSelector(selectPendingEvergreenDepositActions)
+  const txById = useAppSelector(selectTxs)
```
Plural 
```suggestion
  const txs = useAppSelector(selectTxs)
```

gomesalexandre replied:
Going to stamp despite the blocking as this *does not* seem to be a perf bottleneck but still inherently bad for this to run so much

gomesalexandre replied:
Naming looking good now, removal of dep on txs still applicable.

gomesalexandre commented on src/hooks/useActionCenterSubscribers/useEvergreenDepositActionSubscriber.tsx:
```diff
@@ -0,0 +1,79 @@
+
+      const accountAddress = fromAccountId(accountId).account
+      const serializedTxIndex = serializeTxIndex(accountId, stakeTxHash, accountAddress)
+      const transaction = txById[serializedTxIndex]
```
and then remove the dep, this will ensure we don't re-run this guy in a virtually infinite fashion on incoming Txs

```suggestion
      const txsById = selectTxsById(store.getState)
      const maybeTransaction = txById[serializedTxIndex]
```

gomesalexandre replied:
@premiumjibles still applicable

premiumjibles replied:
Noted that makes sense. Will follow up straight after with that change

gomesalexandre reviewed (approved):
LGTM as far as evergreen deposits go 

https://jam.dev/c/0d380aa8-d59c-41da-9429-5270e7f9f8eb

One blocking comment to tackle (and post a jam of the retest) before merging this, all others to tackle at your convenience, feel free to click the merge button once the blocking one is tackled, and if/when non-blocking ones are tackled  🙏🏽

premiumjibles committed:
Clean up type checking

premiumjibles committed:
Update txs name

gomesalexandre reviewed (commented):
Restested locally, still LGTM https://jam.dev/c/394364b6-00be-4a72-98da-f7f187fd9762

One comment for viz TBD in follow-up: https://github.com/shapeshift/web/pull/9873#discussion_r2200115268

Not re-stamping just yet as we'll need to leverage changes from https://github.com/shapeshift/web/pull/9922, notification components are now abstracted

premiumjibles committed:
Convert LP Asset to LP Asset ID

premiumjibles commented on src/state/slices/actionSlice/types.ts:
```diff
@@ -6,13 +6,13 @@ import type { LimitPriceByDirection } from '../limitOrderInputSlice/limitOrderIn
 import type { UnstakingRequest } from '@/pages/RFOX/hooks/useGetUnstakingRequestsQuery/utils'
 
 export enum ActionType {
-  Deposit = 'Deposit',
```
This doesn't seem to be used from what I can see

gomesalexandre replied:
@NeOMakinG confirming we're happy with this removal?

NeOMakinG replied:
Yeah I think this is some dusts from beard initial implementation of the UI, happy to remove

premiumjibles commented:
> One comment for viz TBD in follow-up: [#9873 (comment)](https://github.com/shapeshift/web/pull/9873#discussion_r2200115268)
> 
> Not re-stamping just yet as we'll need to leverage changes from #9922, notification components are now abstracted

Noted for the follow up. Will do that straight after.

I've rebased and reworked to be based off the new components. Retest here

https://jam.dev/c/3b3e17a8-0676-4b5e-b76b-9063a50f5ad7

premiumjibles committed:
Code cleanup adjustments

gomesalexandre reviewed (approved):
Stamping as this does the thing as far as ACC is concerned.

<img width="1156" height="994" alt="image" src="https://github.com/user-attachments/assets/5d924f7e-9f7b-4e41-ac04-bfa726648cc8" />


With that being said and as was noted previously, the ticket only mentions deposits: what about withdraws and claims?

My understanding is, similar to TCY and RFOX tickets similarly missing it, all Tx types should've been handled here.

Did we sync with product about that? Do we plan to add those to the issue or create a new one?
Going to assume that this issue is the same as others and should handle withdraws/claims.

Next steps before merge: 
- change `closes` to `tackle` in the description and do withdraws as a direct follow-up which will close this. No need for additional product spec for withdraws, that's already in the mocks
- claims (claim Txs/claims list) should probably be done in a follow-up of the follow-up since it gets more hairy. It's the same flow as for other bits (Your claim of xyz is available, then claim routes to DeFi modal). Can be checked as a one-liner JSON-RPC call (let's follow the same pattern as others where we check claims once on app load, no need for polling as that would be JSON-RPC spew for realistically no change in claims in the tab lifecycle)

https://github.com/shapeshift/web/blob/cf11486ecfa518a01a8194c8e921318178055602/src/state/slices/opportunitiesSlice/resolvers/ethFoxStaking/index.ts#L154

NeOMakinG commented on src/components/Layout/Header/ActionCenter/components/EvergreenDepositActionCard.tsx:
This is supposed to be a flavor of `GenericTransactionActionCard` instead of duplicating everything, considering we are going to make it a bit more configurable aka accepting a polyglot key as a message with some metadata

I opened an issue for this, so I'm kind of ok to merge as it is if we absolutely take care of https://github.com/shapeshift/web/issues/9970 in a short amount of time

NeOMakinG commented on src/components/Layout/Header/ActionCenter/components/Notifications/EvergreenDepositNotification.tsx:
Same as https://github.com/shapeshift/web/pull/9873/files#r2204466602

NeOMakinG commented on src/components/Layout/Header/ActionCenter/ActionCenter.tsx:
```diff
@@ -89,6 +90,9 @@ export const ActionCenter = memo(() => {
           case ActionType.RfoxClaim: {
             return <RfoxClaimActionCard key={action.id} action={action} />
           }
+          case ActionType.EvergreenDeposit: {
```
Also, if we merge the PR with `EvergreenDeposit` and decide to remove it later, we will need to clear all the actions or migrate the `EvergreenDeposit` actions to `Deposit`, maybe we want to avoid that by doing the generic deposit right now?

premiumjibles replied:
@NeOMakinG agree the above makes sense. This was discussed previously as gomes mentioned. The original plan was to get this in, then follow up with the generalisation when the cards are a bit more configurable. However you're right that we want to avoid any kind of migration here.

Since this PR was raised we've made a number of improvements to the generalisation of the action cards. So it might be worth me taking a step back here and re-looking at what we can do to get this less hyper-specific. And potentially combine it with #9929 and just do this properly from the get go. What are your thoughts? I think the generic cards will still need a bit of fiddling to get this working. For example the default icon for evergreen deposits is a side by side asset icon which won't work with the UI.
